### PR TITLE
Add UnorderedMapInsertOps for coo2crs

### PR DESCRIPTION
**Pull request by e10harvey**
To implement the sparse conversion kernel, coo2crs, I would like to have duplicate key insertions accumulate values in a `Kokkos::UnorderedMap`. This PR adds `UnorderedMapInsertOps::AtomicAdd` and `UnorderedMapInsertOps::NoOp` (the current insertion behavior for duplicate keys).

Related to https://github.com/kokkos/kokkos-kernels/issues/1164.

Going forward, I would also like to extend `Kokkos::UnorderedMap` to support shared memory and add `UnorderedMapInsertOps::Overwrite`.

### CODE REVIEW

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 1]
```diff
@@ -767,6 +810,32 @@ class UnorderedMap {
...
+  bool is_op_noop() const {
+    return std::is_base_of_v<
+        insert_op_type,
+        typename UnorderedMapInsertOpTypes<
+            Kokkos::View<std::remove_const_t<std::conditional_t<
+                             std::is_void_v<Value>, int, Value>> *,
+                         Device>,
+            uint32_t,
+            std::remove_const_t<
+                std::conditional_t<std::is_void_v<Value>, int, Value>>>::NoOp>;
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:823]: That's pretty hard to read 🙂 
```suggestion
    using local_value_type = std::remove_const_t<std::conditional_t<std::is_void_v<Value>, int, Value>>;
    return std::is_base_of_v<
        insert_op_type,
        typename UnorderedMapInsertOpTypes<
            Kokkos::View<local_value_type *, Device>,
            uint32_t, local_value_type>::NoOp>;
```
and similarly below.

File: containers/unit_tests/TestUnorderedMap.hpp [hunk 2]
```diff
@@ -68,7 +82,20 @@ struct TestInsert {
...
+      ++failed_count;
+    else {
+      auto idx = map.find(key);
+      auto ptr = expected_values.data();
+      if (map.is_op_add()) {
+        // Kokkos::abort("adding");
+        Kokkos::atomic_add(&((ptr + idx)[0].v), i);
+      } else if (ret.success() && map.is_op_noop()) {
+        // Kokkos::abort("noop");
+        Kokkos::atomic_store(&((ptr + idx)[0].v), i);
```
Comment (masterleinad) [inline containers/unit_tests/TestUnorderedMap.hpp:96]:
```suggestion
        Kokkos::atomic_add(&((ptr + idx)[0].v), i);
      } else if (ret.success() && map.is_op_noop()) {
        Kokkos::atomic_store(&((ptr + idx)[0].v), i);
```

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 3]
```diff
@@ -767,6 +810,32 @@ class UnorderedMap {
     }
   }
 
+  KOKKOS_FORCEINLINE_FUNCTION
+  bool is_op_noop() const {
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:814]: Do we need these functions at all?

Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:814]: The main need is for asserting in the unit-test. I am open to alternative approaches, perhaps just in the unit-test. Thoughts? I will try to remove these.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 4]
```diff
@@ -175,9 +197,16 @@ class UnorderedMapInsertResult {
...
+          typename Device   = Kokkos::DefaultExecutionSpace,
+          typename Hasher   = pod_hash<std::remove_const_t<Key>>,
+          typename EqualTo  = pod_equal_to<std::remove_const_t<Key>>,
+          typename InsertOp = typename UnorderedMapInsertOpTypes<
+              Kokkos::View<std::remove_const_t<std::conditional_t<
+                               std::is_void_v<Value>, int, Value>> *,
+                           Device>,
+              uint32_t,
+              std::remove_const_t<
+                  std::conditional_t<std::is_void_v<Value>, int, Value>>>::NoOp>
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:209]: You should also document `InsertOp`.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 5]
```diff
@@ -119,6 +119,28 @@ class UnorderedMapInsertResult {
...
 };
 
+/// \class UnorderedMapInsertOpTypes
+/// \brief Operations applied to the values array upon subsequent insertions.
+/// The default behavior when a k,v pair already exists in the UnorderedMap is
+/// to perform no operation. Alternatively, the caller may select to
+/// instantiate the UnorderedMap with the AtomicAdd insert operator such that
+/// duplicate keys accumulate values into the given values array entry.
+template <class ValueTypeView = int, class ValuesIdxType = uint32_t,
+          class ValueType = int>
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:129]: Do you need both `ValueTypeView` and `ValueType`? AFAICT, `impl_value_type` is `typename value_type_view::non_const_value_type` in the implementation of `UnorderedMap`, right?

**Comment by masterleinad** [comment]
Apparently, this fails on the `OpenMP` backend and you will also need to fix the indentation. 🙂

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 5 above)
Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:129]: Thanks for all the helpful feedback, @masterleinad. I think `ValueTypeView` is required since, if the user wants to specify a insert op prior to defining the `map_type`, UnorderedMapInsertOpTypes must be defined before `map_type`. However, perhaps there is a more elegant way to achieve this same result with templates?

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 6]
```diff
@@ -119,6 +119,28 @@ class UnorderedMapInsertResult {
...
+    void op(ValueTypeView, ValuesIdxType, const ValueType *) const {}
+  };
+  struct AtomicAdd {
+    KOKKOS_FUNCTION
+    void op(ValueTypeView values, ValuesIdxType values_idx,
+            const ValueType *v) const {
+      Kokkos::atomic_add(values.data() + values_idx, *v);
+    }
+  };
+};
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:142]: Yes, I agree that we need `ValueTypeView` but I was thinking that you could do 
```suggestion
/// \class UnorderedMapInsertOpTypes
/// \brief Operations applied to the values array upon subsequent insertions.
/// The default behavior when a k,v pair already exists in the UnorderedMap is
/// to perform no operation. Alternatively, the caller may select to
/// instantiate the UnorderedMap with the AtomicAdd insert operator such that
/// duplicate keys accumulate values into the given values array entry.
template <class ValueTypeView = int, class ValuesIdxType = uint32_t>
struct UnorderedMapInsertOpTypes {
  using ValueType = typename ValueTypeView::non_const_value_type;
  struct NoOp {
    KOKKOS_FUNCTION
    void op(ValueTypeView, ValuesIdxType, const ValueType *) const {}
  };
  struct AtomicAdd {
    KOKKOS_FUNCTION
    void op(ValueTypeView values, ValuesIdxType values_idx,
            const ValueType *v) const {
      Kokkos::atomic_add(values.data() + values_idx, *v);
    }
  };
};
```
avoiding `ValueType` to be a template parameter.

Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:142]: Ah, certainly.

**Comment by masterleinad** [comment]
Looks like there is still a segfault (at least for `OpenMP`).

File: containers/unit_tests/TestUnorderedMap.hpp [hunk 7]
```diff
@@ -55,6 +65,15 @@ struct TestInsert {
...
     execution_space().fence();
+
+    if (!rehash_on_fail) {
+      for (unsigned i = 0; i < map.size(); i++) {
+        auto map_idx = expected_values(i).map_idx;
+        if (map_idx != static_cast<unsigned>(~0)) {
+          ASSERT_EQ(expected_values(map_idx).v, map.value_at(map_idx));
+        }
+      }
+    }
```
Comment (masterleinad) [inline containers/unit_tests/TestUnorderedMap.hpp:76]: This is running on the host but the VIew is allocated in the default memory space so our GPU CI fails:
```
 Kokkos::View ERROR: attempt to access inaccessible memory space (label="ExpectedValues")
```

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 8]
```diff
@@ -119,6 +119,36 @@ class UnorderedMapInsertResult {
...
+///
+/// Supported operations:
+///   NoOp:      the first key inserted stores the associated value.
+///   AtomicAdd: duplicate key insertions sum values together.
+template <class ValueTypeView = int, class ValuesIdxType = uint32_t>
+struct UnorderedMapInsertOpTypes {
+  using value_type = typename ValueTypeView::non_const_value_type;
+  struct NoOp {
+    KOKKOS_FUNCTION
+    void op(ValueTypeView, ValuesIdxType, const value_type *) const {}
```
Comment (dalg24) [inline containers/src/Kokkos_UnorderedMap.hpp:141]: Why did you decide to pass a pointer to const rather than passing by value for the last argument?
What types to you expect `value_type` might be?  It is a pessimization for arithmetic types.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 9]
```diff
@@ -119,6 +119,36 @@ class UnorderedMapInsertResult {
...
+/// to perform no operation. Alternatively, the caller may select to
+/// instantiate the UnorderedMap with the AtomicAdd insert operator such that
+/// duplicate keys accumulate values into the given values array entry.
+/// \tparam ValueTypeView The UnorderedMap value array type.
+/// \tparam ValuesIdxType The index type for lookups in the value array.
+///
+/// Supported operations:
+///   NoOp:      the first key inserted stores the associated value.
+///   AtomicAdd: duplicate key insertions sum values together.
+template <class ValueTypeView = int, class ValuesIdxType = uint32_t>
```
Comment (dalg24) [inline containers/src/Kokkos_UnorderedMap.hpp:136]: What does that even mean a default `ValueTypeView` that is `int`?  That would not compile.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 10]
```diff
@@ -296,6 +344,18 @@ class UnorderedMap {
...
+  /// a deeper scope than the original map, the mutable
+  /// m_size member within the original map gets reset
+  /// to 0 when the host mirror object falls out of scope.
+  /// Calling set_flag upon destruction ensures that m_size
+  /// will be recomputed in the parent scope. I think this
+  /// is only a problem when the original map resides on
+  /// the host since some of the host mirror members likely
+  /// point to the same memory locations as the original map
+  /// members.
+  ~UnorderedMap() { set_flag(modified_idx); }
```
Comment (dalg24) [inline containers/src/Kokkos_UnorderedMap.hpp:357]: Is that a drive-by fix?

Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:357]: No? This was triaged first and the addressed with a destructor.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 9 above)
Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:136]: Please suggest an alternative default type.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 8 above)
Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:141]: The signatures of `op` must match.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 9 above)
Comment (dalg24) [reply containers/src/Kokkos_UnorderedMap.hpp:136]: No default type

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 8 above)
Comment (dalg24) [reply containers/src/Kokkos_UnorderedMap.hpp:141]: So?  They both could take a value.  The comment applies to both.

Comment (crtrott) [reply containers/src/Kokkos_UnorderedMap.hpp:141]: Agreed there is no reason these guys take the value as a pointer

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 11]
```diff
@@ -174,10 +204,20 @@ class UnorderedMapInsertResult {
...
-          typename Hasher  = pod_hash<std::remove_const_t<Key>>,
-          typename EqualTo = pod_equal_to<std::remove_const_t<Key>>>
+          typename Device   = Kokkos::DefaultExecutionSpace,
+          typename Hasher   = pod_hash<std::remove_const_t<Key>>,
+          typename EqualTo  = pod_equal_to<std::remove_const_t<Key>>,
+          typename InsertOp = typename UnorderedMapInsertOpTypes<
+              Kokkos::View<std::remove_const_t<std::conditional_t<
+                               std::is_void_v<Value>, int, Value>> *,
+                           Device>,
+              uint32_t>::NoOp>
```
Comment (crtrott) [inline containers/src/Kokkos_UnorderedMap.hpp:220]: Adding a template parameter is a breaking change we can't do that really. We could discuss adding this only if deprecated code off, but we can't simply add it. Alternatively, why can't we have an overload of insert which takes an insert op? Is there a reason this has to be a global template for the class?

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 10 above)
Comment (crtrott) [reply containers/src/Kokkos_UnorderedMap.hpp:357]: We probably want this fix in a separate PR with a test for whatever this fixes.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 12]
```diff
@@ -532,6 +593,7 @@ class UnorderedMap {
         }
 
         result.set_existing(curr, free_existing);
+        if (!is_set) m_insert_op.op(m_values, curr, &v);
```
Comment (crtrott) [inline containers/src/Kokkos_UnorderedMap.hpp:596]: yeah this is the only place where this is used. I think we can just have an overload of insert which takes the insert op. Or maybe even have a different name: insert_accumulate or so. Do you see any issue with that approach?

**Review by crtrott — CHANGES_REQUESTED** [review]
The proposed approach is a breaking change: why can't we just have an insert function overload which takes an op, or add a insert_accumulate function?

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 10 above)
Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:357]: Thanks for the feedback. I'll make another pass soon.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 12 above)
Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:596]: The only issue I see would be a lot of duplicate code.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 11 above)
Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:220]: This seems like the lightest-weight way to add this operation as well as following the implementation style of UnorderedMap. What about keeping the current approach with a single constructor overload, rather than a new insert method, the constructor overload could include the InsertOp type and constructor parameter?

Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:220]: Thinking about this more now, I see no reason for making the InsertOp a global template for the class other than following the existing implementation style of UnorderedMap (e.g. EqualTo, etc.). I will try refactoring with a new insert method instead. Can you elaborate on how the InsertOp template parameter is a breaking change?

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 8 above)
Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:141]: Thinking about this more now, a pointer was used to work around compiler errors when value_type = void for non-insertable UnorderdMaps.

File: containers/src/Kokkos_UnorderedMap.hpp (hunk 11 above)
Comment (e10harvey) [reply containers/src/Kokkos_UnorderedMap.hpp:220]: After trying to remove the InsertOp as a global template parameter, I found that deep copy no longer works. I will look into using insert method overloads rather than a class member.

File: containers/unit_tests/TestUnorderedMap.hpp (hunk 7 above)
Comment (e10harvey) [reply containers/unit_tests/TestUnorderedMap.hpp]: I have been marking feedback I've addressed as resolved. Is the process for the reviewee to mark comments as resolved or should the reviewer do that?

**Comment by e10harvey** [comment]
This is ready for another round of reviews. CI tests for [87c3691](https://github.com/kokkos/kokkos/pull/5877/commits/87c3691558c9c73599455123705257204f9add3d) will fail until #5899 merges.

**Comment by crtrott** [comment]
Needs a rebase. Question at everyone but especially @masterleinad @nmm0 @nliber @dalg24: Do we like the way the insert ops are grouped here mechanically as nested classes in a separate class?

There are alternatives: 
- nested namespace
- make them nested classes in UnorderedMap itself
- named ops (this is what ScatterView does where the ops are called ScatterAdd etc. in the Kokkos::Experimental namespace)

At least we should check for where we do other stuff like that and what approach we take so we can end up with a consistent way of doing this.

**Comment by crtrott** [comment]
Oh just to confirm: the overload approach of insert with a default insert mechanism is good now.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 13]
```diff
@@ -186,6 +217,11 @@ class UnorderedMap {
  public:
   //! \name Public types and constants
   //@{
+  using default_op_type = typename UnorderedMapInsertOpTypes<
```
Comment (nmm0) [inline containers/src/Kokkos_UnorderedMap.hpp:220]: Since you are always passing in `m_values` why not use the type of `m_values` here -- `value_type_view`

**Review by nmm0 — CHANGES_REQUESTED** [review]
Could you add a test for `void` value'd maps (i.e. sets)? What is the expected behavior of the ops when value is `void`?

**Comment by nmm0** [comment]
> > Could you add a test for `void` value'd maps (i.e. sets)? What is the expected behavior of the ops when value is `void`?
> 
> Hi @nmm0. Thanks for your review. The op is not executed when the map is void value'd due to `if(!is_set)`. This is already tested and handled the same way it was before this PR was opened. The specific line of code introduced for getting `void` value'd maps with these insert ops to still build is: https://github.com/kokkos/kokkos/pull/5877/files#diff-2c8acfc1ac5cbbb4cc3da14e466f42c85e415b540c815bf2630c42ec2ac09d59R576.

Thanks for the clarification!

Let's then do:

```cpp
if constexpr (!is_set) {
  arg_insert_op.op(m_values, curr, v);
}
```
Then we don't have to worry about it compiling with a set. Also, in that case, I think we should probably not allow the op to be even specified if inserting into a set.

This is something we probably need to deal with now, because if a user specifies an InsertOp for a set and we clean up the interface later then it would be a breaking change.

I'd like to get other people's opinion on it but one suggestion would be to just do a static_assert, which would be a quick fix rather than making an overload:

```cpp
if constexpr (is_set) {
  static_assert(std::is_same_v<InsertOpType, default_op_type>);
}

**Comment by e10harvey** [comment]
The Jenkins SYCL-OneApi failure is unrelated to these changes:
```
43: [ RUN      ] std_algorithms_sorting_ops_test.is_sorted_until
43: is_sorted_until: dynamic_view, all overloads 
43: /var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsIsSortedUntil.cpp:150: Failure
43: Expected equality of these values:
43:   r3
43:     Which is: 32-byte object <70-0B C3-02 00-00 00-00 80-00 A0-93 9C-7F 00-00 89-8C 01-00 00-00 00-00 F9-02 00-00 00-00 00-00>
43:   gold
43:     Which is: 32-byte object <70-0B C3-02 00-00 00-00 80-00 A0-93 9C-7F 00-00 89-8C 01-00 00-00 00-00 9C-00 00-00 00-00 00-00>
43: /var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsIsSortedUntil.cpp:151: Failure
43: Expected equality of these values:
43:   r4
43:     Which is: 32-byte object <70-0B C3-02 00-00 00-00 80-00 A0-93 9C-7F 00-00 89-8C 01-00 00-00 00-00 01-03 00-00 00-00 00-00>
43:   gold
43:     Which is: 32-byte object <70-0B C3-02 00-00 00-00 80-00 A0-93 9C-7F 00-00 89-8C 01-00 00-00 00-00 9C-00 00-00 00-00 00-00>
43: /var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsIsSortedUntil.cpp:165: Failure
43: Expected equality of these values:
43:   r3
43:     Which is: 32-byte object <70-0B C3-02 00-00 00-00 80-00 A0-93 9C-7F 00-00 89-8C 01-00 00-00 00-00 F9-02 00-00 00-00 00-00>
43:   gold
43:     Which is: 32-byte object <70-0B C3-02 00-00 00-00 80-00 A0-93 9C-7F 00-00 89-8C 01-00 00-00 00-00 9C-00 00-00 00-00 00-00>
43: /var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsIsSortedUntil.cpp:166: Failure
43: Expected equality of these values:
43:   r4
43:     Which is: 32-byte object <70-0B C3-02 00-00 00-00 80-00 A0-93 9C-7F 00-00 89-8C 01-00 00-00 00-00 01-03 00-00 00-00 00-00>
43:   gold
43:     Which is: 32-byte object <70-0B C3-02 00-00 00-00 80-00 A0-93 9C-7F 00-00 89-8C 01-00 00-00 00-00 9C-00 00-00 00-00 00-00>
43: is_sorted_until: stride2_view, all overloads 
43: is_sorted_until: stride3_view, all overloads 
43: [  FAILED  ] std_algorithms_sorting_ops_test.is_sorted_until (72 ms)
```

Full console: https://cloud.cees.ornl.gov/jenkins-ci/blue/rest/organizations/jenkins/pipelines/Kokkos/runs/12424/nodes/49/steps/252/log/?start=0.

**Comment by e10harvey** [comment]
The Jenkins SYCL-OneApi failed again:
```
43: [ RUN      ] std_algorithms_sorting_ops_test.is_sorted_until
43: is_sorted_until: dynamic_view, all overloads 
43: /var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsIsSortedUntil.cpp:148: Failure
43: Expected equality of these values:
43:   r1
43:     Which is: 32-byte object <40-15 04-02 00-00 00-00 80-00 A0-B3 AD-7F 00-00 89-8C 01-00 00-00 00-00 89-8C 01-00 00-00 00-00>
43:   gold
43:     Which is: 32-byte object <40-15 04-02 00-00 00-00 80-00 A0-B3 AD-7F 00-00 89-8C 01-00 00-00 00-00 9C-00 00-00 00-00 00-00>
43: /var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsIsSortedUntil.cpp:163: Failure
43: Expected equality of these values:
43:   r1
43:     Which is: 32-byte object <40-15 04-02 00-00 00-00 80-00 A0-B3 AD-7F 00-00 89-8C 01-00 00-00 00-00 89-8C 01-00 00-00 00-00>
43:   gold
43:     Which is: 32-byte object <40-15 04-02 00-00 00-00 80-00 A0-B3 AD-7F 00-00 89-8C 01-00 00-00 00-00 9C-00 00-00 00-00 00-00>
43: is_sorted_until: stride2_view, all overloads 
43: is_sorted_until: stride3_view, all overloads 
43: [  FAILED  ] std_algorithms_sorting_ops_test.is_sorted_until (143 ms)
```
Can anyone confirm that this failure is unrelated to these changes?

**Comment by masterleinad** [comment]
I would think that the failure is unrelated; the test is not using `Kokkos::UnorderedMap`.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 14]
```diff
@@ -174,6 +204,7 @@ class UnorderedMapInsertResult {
 /// \tparam EqualTo Definition of the equality function for instances of
 ///   <tt>Key</tt>.  The default will do a bitwise equality comparison.
 ///
+///
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:207]: Any reason for this change?
```suggestion
///
```

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 15]
```diff
@@ -442,9 +475,20 @@ class UnorderedMap {
...
+  ///                       Kokkos::UnorderedMapInsertOpTypes for more ops.
+  template <typename InsertOpType = default_op_type>
+  KOKKOS_INLINE_FUNCTION insert_result
+  insert(key_type const &k, impl_value_type const &v = impl_value_type(),
+         InsertOpType arg_insert_op = InsertOpType()) const {
+    if constexpr (is_set) {
+      static_assert(std::is_same_v<InsertOpType, default_op_type>,
+                    "Insert Operations are not supported on sets.");
+      // Silence error #869: parameter "arg_insert_op" was never referenced
+      (void)arg_insert_op;
```
Comment (masterleinad) [inline containers/src/Kokkos_UnorderedMap.hpp:489]:
```suggestion
         [[maybe_unused]] InsertOpType arg_insert_op = InsertOpType()) const {
    if constexpr (is_set) {
      static_assert(std::is_same_v<InsertOpType, default_op_type>,
                    "Insert Operations are not supported on sets.");
```

**Comment by e10harvey** [comment]
I think this is ready to merge. The CI failure appears to be unrelated to these change.

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Comment by nmm0** [comment]
> Needs a rebase. Question at everyone but especially @masterleinad @nmm0 @nliber @dalg24: Do we like the way the insert ops are grouped here mechanically as nested classes in a separate class?
> 
> There are alternatives:
> 
>     * nested namespace
> 
>     * make them nested classes in UnorderedMap itself
> 
>     * named ops (this is what ScatterView does where the ops are called ScatterAdd etc. in the Kokkos::Experimental namespace)
> 
> 
> At least we should check for where we do other stuff like that and what approach we take so we can end up with a consistent way of doing this.

@crtrott did you get satisfactory feedback on this?

**Review by nmm0 — CHANGES_REQUESTED** [review]
I think this is ready depending on what we decide for the convention for Ops

However it needs a changelog entry ([4.1](https://github.com/kokkos/kokkos/issues/5902)) and a corresponding PR for the docs

**Comment by e10harvey** [comment]
> > I think this is ready depending on what we decide for the convention for Ops
> > However it needs a changelog entry ([4.1](https://github.com/kokkos/kokkos/issues/5902)) and a corresponding PR for the docs
> 
> Thinking about Christian's question some more now and in particular how ScatterView handles insertion operators, we should be consistent if possible. @nmm0, @dalg24: Is ScatterView the only other container that supports "operations" in Kokkos core?

It looks like ScatterView has the concept of a ScatterValue type with and specializes on `ScatterValue` for the supported "operation" types like ScatterSum. Thinking about how to support extension of insertion operations in this way is not clear to me. Before this PR, the `Kokkos::UnorderedMap` only requires `operator=` for the given `value_type`. With this PR, we only change what operation is used if the given `key` already exists. So, from my perspective, I think the current approach using `arg_insert_op` is cleaner than introducing the concept of a `ValueType` with a `ValueOp` to `UnorderedMap` since the body of `insert` is not exposed to the user like operations on `ScatterView`. In a later PR, we could revisit `UnorderedMapInsertOpTypes` if they are found to be difficult to use and extend.

**Comment by PhilMiller** [comment]
Retest this please

**Comment by PhilMiller** [comment]
Rerunning the tests, since it looks like Jenkins had a hiccup

**Comment by e10harvey** [comment]
GCC 8.4.0 timed out and the SYCL-OneApi test failed again:
```
43: is_sorted_until: dynamic_view, all overloads 
43: /var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsIsSortedUntil.cpp:149: Failure
43: Expected equality of these values:
43:   r2
43:     Which is: 32-byte object <10-54 8F-02 00-00 00-00 80-00 A0-D3 9A-7F 00-00 89-8C 01-00 00-00 00-00 E1-02 00-00 00-00 00-00>
43:   gold
43:     Which is: 32-byte object <10-54 8F-02 00-00 00-00 80-00 A0-D3 9A-7F 00-00 89-8C 01-00 00-00 00-00 9C-00 00-00 00-00 00-00>
43: /var/jenkins/workspace/Kokkos/algorithms/unit_tests/TestStdAlgorithmsIsSortedUntil.cpp:164: Failure
43: Expected equality of these values:
43:   r2
43:     Which is: 32-byte object <10-54 8F-02 00-00 00-00 80-00 A0-D3 9A-7F 00-00 89-8C 01-00 00-00 00-00 E1-02 00-00 00-00 00-00>
43:   gold
43:     Which is: 32-byte object <10-54 8F-02 00-00 00-00 80-00 A0-D3 9A-7F 00-00 89-8C 01-00 00-00 00-00 9C-00 00-00 00-00 00-00>
43: is_sorted_until: stride2_view, all overloads 
43: is_sorted_until: stride3_view, all overloads 
43: [  FAILED  ] std_algorithms_sorting_ops_test.is_sorted_until (74 ms)
```
